File: dart/lib/src/flag_models.dart
class ApiResponse {
  ApiResponse({required this.flags});

  final List<Flag> flags;

  factory ApiResponse.fromJson(Map<String, dynamic> json) {
    final flagsRaw = json['flags'];
    final flagsList = flagsRaw is List ? flagsRaw : const [];
    return ApiResponse(
      flags: flagsList
          .whereType<Map>()
          .map(
            (raw) => Flag.fromJson(
              raw.map(
                (key, value) => MapEntry(key.toString(), value),
              ),
            ),
          )
          .toList(growable: false),
    );
  }
}

class Flag {
  Flag({
    required this.name,
    required this.enabled,
    required this.strategies,
  });

  final String name;
  final bool enabled;
  final List<Strategy> strategies;

  factory Flag.fromJson(Map<String, dynamic> json) {
    final strategiesRaw = json['strategies'];
    final strategiesList = strategiesRaw is List ? strategiesRaw : const [];
    return Flag(
      name: json['name'] is String ? json['name'] as String : '',
      enabled: json['enabled'] is bool ? json['enabled'] as bool : false,
      strategies: strategiesList
          .whereType<Map>()
          .map(
            (raw) => Strategy.fromJson(
              raw.map(
                (key, value) => MapEntry(key.toString(), value),
              ),
            ),
          )
          .toList(growable: false),
    );
  }
}

class Strategy {
  Strategy({
    required this.name,
    required this.parameters,
    required this.constraints,
  });

  final String name;
  final Map<String, dynamic> parameters;
  final List<Constraint> constraints;

  factory Strategy.fromJson(Map<String, dynamic> json) {
    final constraintsRaw = json['constraints'];
    final constraintsList = constraintsRaw is List ? constraintsRaw : const [];
    final parametersRaw = json['parameters'];
    return Strategy(
      name: json['name'] is String ? json['name'] as String : '',
      parameters: parametersRaw is Map
          ? parametersRaw.map(
              (key, value) => MapEntry(key.toString(), value),
            )
          : <String, dynamic>{},
      constraints: constraintsList
          .whereType<Map>()
          .map(
            (raw) => Constraint.fromJson(
              raw.map(
                (key, value) => MapEntry(key.toString(), value),
              ),
            ),
          )
          .toList(growable: false),
    );
  }
}

class Constraint {
  Constraint({
    required this.contextName,
    required this.operator,
    required this.values,
    required this.inverted,
    required this.caseInsensitive,
  });

  final String contextName;
  final String operator;
  final List<String> values;
  final bool inverted;
  final bool caseInsensitive;

  factory Constraint.fromJson(Map<String, dynamic> json) {
    final valuesRaw = json['values'];
    final valuesList = valuesRaw is List ? valuesRaw : const [];
    return Constraint(
      contextName: json['context_name'] is String
          ? json['context_name'] as String
          : '',
      operator: json['operator'] is String ? json['operator'] as String : '',
      values: valuesList.map((value) => value.toString()).toList(growable: false),
      inverted: json['inverted'] is bool ? json['inverted'] as bool : false,
      caseInsensitive: json['case_insensitive'] is bool
          ? json['case_insensitive'] as bool
          : false,
    );
  }
}

File: elixir/lib/bandeira/flag_models.ex
defmodule Bandeira.FlagModels do
  @moduledoc false

  defmodule Flag do
    @moduledoc false
    defstruct name: "", enabled: false, strategies: []

    @type t :: %__MODULE__{
            name: String.t(),
            enabled: boolean(),
            strategies: [Bandeira.FlagModels.Strategy.t()]
          }
  end

  defmodule Strategy do
    @moduledoc false
    defstruct name: "", parameters: %{}, constraints: []

    @type t :: %__MODULE__{
            name: String.t(),
            parameters: map(),
            constraints: [Bandeira.FlagModels.Constraint.t()]
          }
  end

  defmodule Constraint do
    @moduledoc false
    defstruct context_name: "", operator: "", values: [], inverted: false, case_insensitive: false

    @type t :: %__MODULE__{
            context_name: String.t(),
            operator: String.t(),
            values: [String.t()],
            inverted: boolean(),
            case_insensitive: boolean()
          }
  end

  alias __MODULE__.{Constraint, Flag, Strategy}

  @spec parse_response(map()) :: %{optional(String.t()) => Flag.t()}
  def parse_response(data) when is_map(data) do
    data
    |> get(:flags, [])
    |> case do
      flags when is_list(flags) ->
        Enum.reduce(flags, %{}, fn raw, acc ->
          flag = to_flag(raw)

          if flag.name == "" do
            acc
          else
            Map.put(acc, flag.name, flag)
          end
        end)

      _ ->
        %{}
    end
  end

  def parse_response(_), do: %{}

  @spec to_flag(map()) :: Flag.t()
  def to_flag(raw) when is_map(raw) do
    %Flag{
      name: get(raw, :name, "") |> to_string_safe(),
      enabled: get(raw, :enabled, false) == true,
      strategies: to_strategies(get(raw, :strategies, []))
    }
  end

  def to_flag(_), do: %Flag{}

  defp to_strategies(strategies) when is_list(strategies) do
    Enum.map(strategies, &to_strategy/1)
  end

  defp to_strategies(_), do: []

  defp to_strategy(raw) when is_map(raw) do
    %Strategy{
      name: get(raw, :name, "") |> to_string_safe(),
      parameters: to_parameters(get(raw, :parameters, %{})),
      constraints: to_constraints(get(raw, :constraints, []))
    }
  end

  defp to_strategy(_), do: %Strategy{}

  defp to_parameters(parameters) when is_map(parameters), do: parameters
  defp to_parameters(_), do: %{}

  defp to_constraints(constraints) when is_list(constraints) do
    Enum.map(constraints, &to_constraint/1)
  end

  defp to_constraints(_), do: []

  defp to_constraint(raw) when is_map(raw) do
    %Constraint{
      context_name: get(raw, :context_name, "") |> to_string_safe(),
      operator: get(raw, :operator, "") |> to_string_safe(),
      values: to_values(get(raw, :values, [])),
      inverted: get(raw, :inverted, false) == true,
      case_insensitive: get(raw, :case_insensitive, false) == true
    }
  end

  defp to_constraint(_), do: %Constraint{}

  defp to_values(values) when is_list(values), do: Enum.map(values, &to_string_safe/1)
  defp to_values(_), do: []

  defp get(map, key, default) do
    cond do
      Map.has_key?(map, key) -> Map.get(map, key)
      is_atom(key) and Map.has_key?(map, Atom.to_string(key)) -> Map.get(map, Atom.to_string(key))
      true -> default
    end
  end

  defp to_string_safe(nil), do: ""
  defp to_string_safe(value), do: to_string(value)
end
